controller/lib/hal/stepper.cpp

```diff
@@ -0,0 +1,820 @@
+
+// This array holds the length of each parameter in units of
+// bytes, rounded up to the nearest byte.  This info is based
+// on table 12 in the powerSTEP chip data sheet
+uint8_t StepMotor::param_len_[32] = {
+    0, // 0x00 - No valid parameter with ID 0
+    3, // 0x01 - Absolute position (22 bits)
+    2, // 0x02 - Electrical position (9 bits)
+    3, // 0x03 - Mark position (22 bits)
+    3, // 0x04 - Current speed (20 bits)
+    2, // 0x05 - Acceleration (12 bits)
+    2, // 0x06 - Deceleration (12 bits)
+    2, // 0x07 - Maximum speed (10 bits)
+    2, // 0x08 - Minimum speed (12 bits)
+    1, // 0x09 - KVAL_HOLD Holding K VAL (8 bits)
+    1, // 0x0A - KVAL_RUN Constant speed K VAL (8 bits)
+    1, // 0x0B - KVAL_ACC Acceleration starting K VAL (8 bits)
+    1, // 0x0C - KVAL_DEC Deceleration starting K VAL (8 bits)
+    2, // 0x0D - INT_SPEED Intersect speed (14 bits)
+    1, // 0x0E - ST_SLP Start slope (8 bits)
+    1, // 0x0F - FN_SLP_ACC Acceleration final slope (8 bits)
+    1, // 0x10 - FN_SLP_DEC Deceleration final slope (8 bits)
+    1, // 0x11 - K_THERM Thermal compensation factor (4 bits)
+    1, // 0x12 - ADC output (5 bits)
+    1, // 0x13 - OCD threshold (5 bits)
+    1, // 0x14 - STALL_TH STALL threshold (5 bits)
+    2, // 0x15 - Full-step speed (11 bits)
+    1, // 0x16 - Step mode (8 bits)
+    1, // 0x17 - Alarm enables (8 bits)
+    2, // 0x18 - Gate driver configuration (11 bits)
+    1, // 0x19 - Gate driver configuration (8 bits)
+    2, // 0x1A - IC configuration (16 bits)
+    2, // 0x1B - Status (16 bits)
+    0, // 0x1C - No such parameter
+    0, // 0x1D - No such parameter
+    0, // 0x1E - No such parameter
+    0, // 0x1F - No such parameter
+};
+
+// These constants are used to convert speeds in steps/sec
+// to the weird values used by the stepper driver chip.
+// Note that different registers use different conversion
+// factors.  See the chip data sheet for details.
+static const float tick_time = 250e-9f;
+static const float cvt_vel_crnt_speed_reg = tick_time * (1 << 28);
+static const float cvt_vel_max_speed_reg = tick_time * (1 << 18);
+static const float cvt_vel_min_speed_reg = tick_time * (1 << 24);
+static const float cvt_vel_fs_speed_reg = tick_time * (1 << 18);
+static const float cvt_vel_int_speed_reg = tick_time * (1 << 26);
+
+// The number of microsteps / full step.
+// For now this is a constant, we're just using the
+// default value of the chip.
+static const int ustep_per_step_ = 128;
+
+// These functions raise and lower the chip select pin
+static inline void CS_High() { GPIO_SetPin(GPIO_B_BASE, 6); }
+static inline void CS_Low() { GPIO_ClrPin(GPIO_B_BASE, 6); }
+
+StepMotor::StepMotor() {
+  queue_count_ = 0;
+  queue_ndx_ = 0;
+  cmd_ptr_ = 0;
+  cmd_remain_ = 0;
+  save_response_ = false;
+  steps_per_rev_ = 200;
+}
+
+StepMtrErr StepMotor::SetParam(StepMtrParam param, uint32_t value) {
+  uint8_t P = static_cast<uint8_t>(param);
```

Comment:
If StepMtrParam must fit within uint8_t, it might be worth annotating the enum.

```
// This enum must fit within one byte because XXX
enum class StepMtrParam : uint8_t
```